xen/drivers/passthrough/arm/ipmmu-vmsa.c

```diff
@@ -1411,14 +1520,47 @@ static int ipmmu_add_device(u8 devfn, struct device *dev)
     if ( dev_is_pci(dev) )
     {
         struct pci_dev *pdev = dev_to_pci(dev);
+        struct gen4_pci_ipmmu_info *info;
+        struct dt_device_node *np;
+        unsigned int reg_id, osid;
         int ret;
 
-        /* TODO: this check needs to be removed once PCI OSIDs are in place. */
-        if ( pdev->sbdf.sbdf != PCI_BDF(0, 0, 0) )
+        np = pci_find_host_bridge_node(dev);
+        if ( !np )
+            return -ENODEV;
+
+        if ( dt_node_path_is_equal(np, "/soc/pcie@e65d0000") )
+            info = &pci_ipmmu_info[0];
+        else if ( dt_node_path_is_equal(np, "/soc/pcie@e65d8000") )
+            info = &pci_ipmmu_info[1];
+        else
+            return -ENODEV;
+
+        if ( fwspec->num_ids != 1 || fwspec->ids[0] < info->utlb_osid0 ||
+             fwspec->ids[0] >= (info->utlb_osid0 + NUM_OSID_REGS) )
+            return -EINVAL;
+
+        osid_init_regs(info);
+
+        ret = osid_alloc_reg(info);
+        if ( ret < 0 )
         {
-            ret = ipmmu_assign_device(pdev->domain, devfn, dev, 0);
-            if (ret)
-                return ret;
+            dev_err(dev, "No unused OSID regs\n");
+            return ret;
+        }
+        reg_id = ret;
+
+        osid = fwspec->ids[0] - info->utlb_osid0;
+        osid_bdf_set(info, reg_id, osid, pdev->sbdf.bdf);
+        bdf_msk_set(info, reg_id, 0);
```

Comment:
either osid_bdf_set and mask_bdf_set or other way around, e.g. bdf_osid and bdf_mask

Reply:
I would prefer to leave **bdf_msk_set** as is. The function was named in a such way because **bdf_msk** is a name of the corresponding bit field which function sets. In turn **osid_bdf_set** sets two separate bit fields **osid** and **bdf**

Reply:
makes sense